DynamicTablesPkg/Library/Acpi/Common/AcpiDbg2Lib/Dbg2Generator.c

```diff
@@ -266,33 +172,166 @@ FreeDbg2TableEx (
+  IN  CONST EDKII_CONFIGURATION_MANAGER_PROTOCOL   *CONST  CfgMgrProtocol,
+  IN OUT EFI_ACPI_DEBUG_PORT_2_DESCRIPTION_TABLE           **AcpiDbg2,
+  IN OUT UINT32                                            *AcpiDbg2Len,
+  IN CM_ARCH_COMMON_DBG2_DEVICE_INFO                       *DeviceInfo,
+  IN CM_ARCH_COMMON_MEMORY_RANGE_DESCRIPTOR                *MemoryRange OPTIONAL
+  )
+{
+  EFI_STATUS                                     Status;
+  UINTN                                          Index;
+  CM_ARCH_COMMON_MEMORY_RANGE_DESCRIPTOR         *MemoryRanges;
+  UINT32                                         MemoryRangesCount;
+  UINT16                                         Dbg2DeviceSize;
+  EFI_ACPI_DBG2_DEBUG_DEVICE_INFORMATION_STRUCT  *Dbg2Device;
+  VOID                                           *NewAcpiDbg2;
+  EFI_ACPI_6_3_GENERIC_ADDRESS_STRUCTURE         *BaseAddressRegister;
+  UINT32                                         *AddressSize;
+
+  // Parameter validation
+  if ((CfgMgrProtocol == NULL) || (AcpiDbg2 == NULL) ||
+      (AcpiDbg2Len == NULL) || (DeviceInfo == NULL))
+  {
+    DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Invalid parameters to PopulateDbg2Device\n"));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  // Check string length before concatenation
+  if ((DeviceInfo->ObjectName[0] != '\0') &&
+      (AsciiStrLen (SB_SCOPE) + AsciiStrLen (DeviceInfo->ObjectName) >= DBG2_NAMESPACESTRING_FIELD_SIZE))
+  {
+    DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Namespace string too long\n"));
+    return EFI_BUFFER_TOO_SMALL;
+  }
+
+  if (DeviceInfo->AddressResourceToken != CM_NULL_TOKEN) {
+    Status = GetEArchCommonObjMemoryRangeDescriptor (
+               CfgMgrProtocol,
+               DeviceInfo->AddressResourceToken,
+               &MemoryRanges,
+               &MemoryRangesCount
+               );
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Failed to get memory range descriptor. Status = %r\n", Status));
+      return Status;
+    }
+
+    if (MemoryRangesCount > MAX_UINT8) {
+      DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Too many memory ranges. Count = %u\n", MemoryRangesCount));
+      return EFI_INVALID_PARAMETER;
+    }
+
+    // Validate all memory ranges
+    for (Index = 0; Index < MemoryRangesCount; Index++) {
+      if (MemoryRanges[Index].Length > MAX_UINT32) {
+        DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Memory range length too large. Length = %u\n", MemoryRanges[Index].Length));
+        return EFI_INVALID_PARAMETER;
+      }
+    }
+  } else {
+    MemoryRanges      = MemoryRange;
+    MemoryRangesCount = 1;
+  }
+
+  Dbg2DeviceSize = (UINT16)(sizeof (EFI_ACPI_DBG2_DEBUG_DEVICE_INFORMATION_STRUCT) +
+                            ((sizeof (EFI_ACPI_6_3_GENERIC_ADDRESS_STRUCTURE) + sizeof (UINT32)) * (MemoryRangesCount)) +
+                            (sizeof (CHAR8) * DBG2_NAMESPACESTRING_FIELD_SIZE));
+
+  NewAcpiDbg2 = ReallocatePool (*AcpiDbg2Len, *AcpiDbg2Len + Dbg2DeviceSize, *AcpiDbg2);

```

Comment:
I think it would be simpler to compute the overall DBG2 table size first and do one allocation rather than re-allocating the table N times.

It should be possible to save the MemoryRangeDescriptors in an array to avoid fetching them multiple times